homeassistant/components/sensor/linky.py

```diff
@@ -5,88 +5,227 @@
 })
 
 
 def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Configure the platform and add the Linky sensor."""
-    username = config[CONF_USERNAME]
-    password = config[CONF_PASSWORD]
-    timeout = config[CONF_TIMEOUT]
+    """Set up the Linky sensor."""
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    timeout = config.get(CONF_TIMEOUT)
+    name = config.get(CONF_NAME)
+    peak_hours = config.get(PEAK_HOURS)
+    peak_hours_cost = config.get(PEAK_HOURS_COST)
+    offpeak_hours_cost = config.get(OFFPEAK_HOURS_COST)
 
-    from pylinky.client import LinkyClient, PyLinkyError
-    client = LinkyClient(username, password, None, timeout)
-    try:
-        client.fetch_data()
-    except PyLinkyError as exp:
-        _LOGGER.error(exp)
-        client.close_session()
-        return
+    linky_data = LinkyData(username, password, timeout)
+    linky_data.update()
 
-    devices = [LinkySensor('Linky', client)]
-    add_entities(devices, True)
+    add_entities([LinkySensor(name, linky_data, peak_hours,
+                              peak_hours_cost, offpeak_hours_cost)])
+    return True
 
 
 class LinkySensor(Entity):
     """Representation of a sensor entity for Linky."""
 
-    def __init__(self, name, client):
+    def __init__(self, name, linky_data,
+                 peak_hours, peak_hours_cost,
+                 offpeak_hours_cost):
         """Initialize the sensor."""
         self._name = name
-        self._client = client
+        self._peak_hours = peak_hours
+        self._peak_hours_cost = peak_hours_cost
+        self._offpeak_hours_cost = offpeak_hours_cost
+        self._unit = KILOWATT_HOUR
+        self._icon = "mdi:flash"
+        self._lk = linky_data
         self._state = None
+        self._attributes = {}
+        self.update()
 
     @property
     def name(self):
         """Return the name of the sensor."""
         return self._name
 
+    @property
+    def unit_of_measurement(self):
+        """Return the unit the value is expressed in."""
+        return self._unit
+
+    @property
+    def state_attributes(self):
+        """Return the state attributes."""
+        return self._attributes
+
     @property
     def state(self):
         """Return the state of the sensor."""
         return self._state
 
     @property
-    def unit_of_measurement(self):
-        """Return the unit of measurement."""
-        return 'kWh'
+    def icon(self):
+        """Return the icon of the sensor."""
+        return self._icon
 
     @Throttle(SCAN_INTERVAL)
     def update(self):
         """Fetch new state data for the sensor."""
+        _LOGGER.debug("Start of update of linky data")
+        self._lk.update()
+        if not self._lk.success:
+            self._state = STATE_UNAVAILABLE
+            self._attributes = []
+            return
+
+        self._state = self._lk.daily[1][CONSUMPTION]
+        self._attributes["halfhourly"] = [d[CONSUMPTION]
```

Comment:
We shouldn't do these calculations as part of the integration, instead it should be made a standalone integration, like the previous mentioned utility meter.